src/operator/tensor/broadcast_reduce_op.h

```diff
@@ -1147,16 +1249,26 @@ inline void BroadcastComputeImpl(const nnvm::NodeAttrs& attrs,
           outputs[0].get_with_shape<xpu, 2, OType>(dst_shape.get<2>(), s);
         Tensor<xpu, 2, IType> data =
           inputs[0].get_with_shape<xpu, 2, IType>(src_shape.get<2>(), s);
-        Kernel<broadcast_kernel<mshadow_op::identity>, xpu>::Launch(
-          s, out.shape_.Size(), data.dptr_, out.dptr_, aux_data, req[0], 2);
+        if (isCPU) {
+          Kernel<broadcast_kernel_cpu<mshadow_op::identity>, xpu>::Launch(
+            s, data.shape_.Size(), data.dptr_, out.dptr_, aux_data, req[0], 2);
+        } else {
+          Kernel<broadcast_kernel<mshadow_op::identity>, xpu>::Launch(
+            s, out.shape_.Size(), data.dptr_, out.dptr_, aux_data, req[0], 2);
+        }
       } else {
         const int ndim = MXNET_SPECIAL_MAX_NDIM;
         Tensor<xpu, ndim, OType> out =
           outputs[0].get_with_shape<xpu, ndim, OType>(dst_shape.get<ndim>(), s);
         Tensor<xpu, ndim, IType> data =
           inputs[0].get_with_shape<xpu, ndim, IType>(src_shape.get<ndim>(), s);
-        Kernel<broadcast_kernel<mshadow_op::identity>, xpu>::Launch(
-          s, out.shape_.Size(), data.dptr_, out.dptr_, aux_data, req[0], ndim);
+        if (isCPU) {
+          Kernel<broadcast_kernel_cpu<mshadow_op::identity>, xpu>::Launch(
```

Comment:
If it's only CPU you don't need to use the generic kernel, right? You can just use the Kernel<OP, cpu> template.

Reply:
Kernel<broadcast_kernel_cpu<mshadow_op::identity>, cpu> actually causes build failure and gives the following error:
```
../src/operator/nn/./../tensor/broadcast_reduce_op.h(1246): error: no instance of overloaded function "mxnet::op::mxnet_op::Kernel<OP, mxnet::cpu>::Launch [with OP=mxnet::op::broadcast_kernel_cpu<mxnet::op::mshadow_op::identity>]" matches the argument list
            argument types are: (mshadow::Stream<mshadow::gpu> *, mshadow::index_t, float *, float *, mxnet::op::<unnamed>::ShapeAndStride, const mxnet::OpReqType, int)
          detected during:
            instantiation of "void mxnet::op::BroadcastComputeImpl<xpu>(const nnvm::NodeAttrs &, const mxnet::OpContext &, const std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob>> &, const std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType>> &, const std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob>> &, const mxnet::TShape &) [with xpu=mxnet::gpu]"
../src/operator/nn/./moments-inl.h(215): here
            instantiation of "void mxnet::op::MomentsBackwardImpl<xpu>(const nnvm::NodeAttrs &, const mxnet::OpContext &, const std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob>> &, const std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType>> &, const std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob>> &, const dmlc::optional<mxnet::TShape> &) [with xpu=mxnet::gpu]"
../src/operator/nn/./moments-inl.h(249): here
            instantiation of "void mxnet::op::MomentsBackward<xpu>(const nnvm::NodeAttrs &, const mxnet::OpContext &, const std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob>> &, const std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType>> &, const std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob>> &) [with xpu=mxnet::gpu]"
../src/operator/nn/moments.cu(36): here
```